File: cpp/binary_tree.cpp
#include <memory>
#include <vector>
#include <iostream>

struct node {
    int value;
    std::unique_ptr<node> children[2];
    node(int value) 
        : value{value}, children{nullptr, nullptr}
    {

    } 
};

struct binary_tree { 
    std::unique_ptr<node> root;
    binary_tree() : root{nullptr} 
    {}

    std::unique_ptr<node> recursive_insert(node* parent, int value) {
        if (parent == nullptr) {
            return std::make_unique<node>(value);
        } 
        auto idx = (value < parent->value) ? 0 : 1;
        auto curr_child = parent->children[idx].get();  
        auto new_node = recursive_insert(curr_child, value);
        if (curr_child == nullptr) {
            parent->children[idx] = std::move(new_node);
        }
        return nullptr;
    }

    void insert(int value) {
        if (root == nullptr){
            root = std::make_unique<node>(value);
            return;
        }
        recursive_insert(root.get(), value);
    }
    void recursive_print(node* ptr, int level) {
        if(ptr) {
            recursive_print(ptr->children[1].get(), level + 1);
            for (size_t i = 0; i < level; i++){
                std::cout << "    ";
            }
            std::cout << ptr->value << std::endl;
            recursive_print(ptr->children[0].get(), level + 1);
        }
    }

    void print() {
        recursive_print(root.get(), 0);        
    }
};

int main() {
    binary_tree b;
    b.insert(5);
    b.insert(3);
    b.insert(7);
    b.insert(2);
    b.insert(4);
    b.insert(6);
    b.insert(9);
    b.print();
    return 0;
}

File: rust/src/binary_tree.rs

struct Node {
    value:i32,
    children: [Option<Box<Node>>; 2]
}

struct BinaryTree{
    root: Option<Box<Node>>
}

fn _recursive_insert(ptr: &mut Option<Box<Node>>, value:i32) -> Option<Box<Node>> {
    if ptr.is_none() {
        return Some(Box::new(Node { value, children: [None, None] }));
    }
    if let Some(parent) = ptr {
        let idx = if value < parent.value { 0 } else { 1 };
        let ret = _recursive_insert(&mut parent.children[idx], value);
        if parent.children[idx].is_none() {
            parent.children[idx] = ret;
        }
    }
    None
}

// doc ref
// https://doc.rust-lang.org/rust-by-example/flow_control/for.html
fn _recursive_print(parent: &Option<Box<Node>>, level:i32){
    if let Some(ptr) = parent {
        _recursive_print(&ptr.children[1], level + 1);
        for _ in 0..level {
            print!("    ");
        }
        println!("{}", ptr.value);
        _recursive_print(&ptr.children[0], level + 1);
    }
}
impl BinaryTree {
    fn new() -> Self {
        BinaryTree{
            root: None
        }
    }
    fn insert(&mut self, value:i32){
        if self.root.is_none() {
            self.root = Some(Box::new(Node { value, children: [None, None] }));
        } else {
            _recursive_insert(&mut self.root, value);
        }
    }
    fn print(&mut self) {
        _recursive_print(&self.root, 0);
    }
}

fn main() {
    let mut b  = BinaryTree::new();
    b.insert(5);
    b.insert(3);
    b.insert(7);
    b.insert(2);
    b.insert(4);
    b.insert(6);
    b.insert(9);
    b.print();
}